src/context.ts

```diff
@@ -3,9 +3,130 @@ import { GlobalLogger as Logger, Logger as DBOSLogger } from "./telemetry/logs";
+  span?: Span;
+  authenticatedUser?: string;
+  authenticatedRoles?: string[];
+  assumedRole?: string;
+  request?: HTTPRequest;
+}
+
+function isWithinWorkflowCtx(ctx: DBOSLocalCtx) {
+  if (ctx.workflowId === undefined) return false;
+  return true;
+}
+
+function isInStepCtx(ctx: DBOSLocalCtx) {
+  if (ctx.workflowId === undefined) return false;
+  if (ctx.curStepFunctionId) return true;
+  return false;
+}
+
+function isInTxnCtx(ctx: DBOSLocalCtx) {
+  if (ctx.workflowId === undefined) return false;
+  if (ctx.curTxFunctionId) return true;
+  return false;
+}
+
+function isInWorkflowCtx(ctx: DBOSLocalCtx) {
+  if (!isWithinWorkflowCtx(ctx)) return false;
+  if (isInStepCtx(ctx)) return false;
+  if (isInTxnCtx(ctx)) return false;
+  return true;
+}
+
+export const asyncLocalCtx = new AsyncLocalStorage<DBOSLocalCtx>();
+
+export function getCurrentContextStore() : DBOSLocalCtx | undefined {
+  return asyncLocalCtx.getStore();
+}
+
+export function getCurrentDBOSContext() : DBOSContext | undefined {
+  return asyncLocalCtx.getStore()?.ctx;
+}
+
+export function assertCurrentDBOSContext(): DBOSContext {
+  const ctx = asyncLocalCtx.getStore()?.ctx;
+  if (!ctx) throw new DBOSNotRegisteredError("No current DBOS Context");
+  return ctx;
+}
+
+export function assertCurrentWorkflowContext(): WorkflowContextImpl {
+  const ctx = assertCurrentDBOSContext();
+  if (!isInWorkflowCtx(asyncLocalCtx.getStore()!)) {
+    throw new DBOSInvalidWorkflowTransitionError();
+  }
+  return ctx as WorkflowContextImpl;
+}
+
+export async function runWithDBOSContext<R>(ctx: DBOSContext, callback: ()=>Promise<R>) {
+  return await asyncLocalCtx.run({
+    ctx,
+    workflowId: ctx.workflowUUID,
+  }, callback);
+}
+
+export async function runWithTopContext<R>(ctx: DBOSLocalCtx, callback: ()=>Promise<R>): Promise<R> {
+  return await asyncLocalCtx.run(ctx, callback);
+}
+
+export async function runWithTransactionContext<Client extends UserDatabaseClient, R>(ctx: TransactionContextImpl<Client>, callback: ()=>Promise<R>) {
+  // Check we are in a workflow context and not in a step / transaction already
+  const pctx = getCurrentContextStore();
+  if (!pctx) throw new DBOSInvalidWorkflowTransitionError();
+  if (!isInWorkflowCtx(pctx)) throw new DBOSInvalidWorkflowTransitionError();
+  return await asyncLocalCtx.run({
+    ctx,
+    workflowId: ctx.workflowUUID,
+    curTxFunctionId: ctx.functionID,
+    parentCtx: pctx,
+  },
+  callback);
+}
+
+export async function runWithStepContext<R>(ctx: StepContextImpl, callback: ()=>Promise<R>) {
+  // Check we are in a workflow context and not in a step / transaction already
+  const pctx = getCurrentContextStore();
+  if (!pctx) throw new DBOSInvalidWorkflowTransitionError();
+  if (!isInWorkflowCtx(pctx)) throw new DBOSInvalidWorkflowTransitionError();
+
+  return await asyncLocalCtx.run({
+    ctx,
+    workflowId: ctx.workflowUUID,
+    curStepFunctionId: ctx.functionID,
+    parentCtx: pctx,
+  },
+  callback);
+}
```

Comment:
Who are these "run with X context" functions intended for?

Reply:
Intended for?  Or used for?  That one is used to run the callback function within availability of the StepContext.  Since we know it is a step context (and we're invoking a step), we can pick things out of it that may not have been possible w/ the base interface.